src/cli.ts

```diff
@@ -25,35 +39,120 @@ async function main(argv: minimist.ParsedArgs) {
+    throw new ReferenceError(
+      `You have specified a single file ${argOut} output for a multi file input ${argIn}. This feature is not yet supported, refer to issue #272 (https://github.com/bcherny/json-schema-to-typescript/issues/272)`
+    )
+  }
 
   try {
-    const schema: JSONSchema4 = JSON.parse(await readInput(argIn))
-    const ts = await compile(schema, argIn, argv as Partial<Options>)
-    await writeOutput(ts, argOut)
+    // Process input as either glob, directory, or single file
+    if (ISGLOB) {
+      await processGlob(argIn, argOut, argv as Partial<Options>)
+    } else if (ISDIR) {
+      await processDir(argIn, argOut, argv as Partial<Options>)
+    } else {
+      await processFile(argIn, argOut, argv as Partial<Options>)
+    }
   } catch (e) {
     console.error(whiteBright.bgRedBright('error'), e)
     process.exit(1)
   }
 }
 
-function readInput(argIn?: string) {
-  if (!argIn) {
-    return new Promise(stdin)
+// check if path is an existing directory
+function isDir(path: string): boolean {
+  return existsSync(path) && lstatSync(path).isDirectory()
+}
+
+async function processGlob(argIn: string, argOut: string | undefined, argv: Partial<Options>) {
+  const files = await glob(argIn) // execute glob pattern match
+
+  if (files.length === 0) {
+    throw ReferenceError(
+      `You passed a glob pattern "${argIn}", but there are no files that match that pattern in ${process.cwd()}`
+    )
   }
-  return readFile(resolve(process.cwd(), argIn), 'utf-8')
+  // create output directory if it does not exist
+  if (argOut && !existsSync(argOut)) {
+    await mkdirp(argOut)
+  }
+
+  return Promise.all(
+    files.map(file => {
+      const outPath = argOut ? `${argOut}/${basename(file, '.json')}.d.ts` : argOut
+      processFile(file, outPath, argv)
+    })
+  )
+}
+
+async function processDir(argIn: string, argOut: string | undefined, argv: Partial<Options>): Promise<void[]> {
+  const files = getPaths(argIn)
+
+  return Promise.all(
+    files.map(file => {
+      if (!argOut) {
+        processFile(file, argOut, argv)
+      } else {
+        /*
+        the following logic determines the out path by comparing the in path to the users specified out path.
+        For example, if input directory MultiSchema looks like:
+        MultiSchema/foo/a.json
+        MultiSchema/bar/fuzz/c.json
+        MultiSchema/bar/d.json
+        And the user wants the outputs to be in MultiSchema/Out, then this code will be able to map the inner directories foo, bar, and fuzz into the intended Out directory.
+        */
+        const outPathList = argOut.split('/')
+        const inPathList = file.split('/')
+
+        const intersection = outPathList.filter(x => inPathList.includes(x))
+        const difference = outPathList
+          .filter(x => !inPathList.includes(x))
+          .concat(inPathList.filter(x => !outPathList.includes(x)))
+
+        let outPath = join(...intersection, ...difference)
```

Comment:
Would you mind pulling lines 110-118 out into its own function, and writing some unit tests for it? I'm having trouble understanding if this will work right in all cases.

Reply:
It uses set theory. Not sure if unit testing would be best here, but happy to provide some if that is what you think is best.

Reply:
The _proof_ for this is:

The first two lines split the file path into lists separated by the `/`.
```
const outPathList = o.split('/')
const inPathList = i.split('/')
```

The next line creates an intersection set of the two lists. This will contain only the parts of the file path that are shared. 
```
const intersection = outPathList.filter(x => inPathList.includes(x))
```

Next, we create a symmetric difference of the out path. This is essentially the list of paths that are not shared between the two paths.
```
  const difference = outPathList
    .filter(x => !inPathList.includes(x))
    .concat(inPathList.filter(x => !outPathList.includes(x)))
```

Then we join these two lists together (with the intersection list first) to produce the transformed output.
```
return join(...intersection, ...difference)
```
---
An example, consider the following input directory `MultiSchema`
```
MultiSchema/foo/a.json
MultiSchema/bar/fuzz/c.json
MultiSchema/bar/d.json
```
And the desired output directory is `MultiSchema/Out`
Then the CLI will perform the following function calls:
```
pathTransform('MultiSchema/Out', 'MultiSchema/foo/a.json')
pathTransform('MultiSchema/Out', 'MultiSchema/bar/fuzz/c.json')
pathTransform('MultiSchema/Out', 'MultiSchema/bar/d.json')
```
And will result in the intended output:
```
'MultiSchema/Out/foo/a.json'
'MultiSchema/Out/bar/d.json'
'MultiSchema/Out/bar/fuzz/c.json'
```

Reply:
A proof is great, tests are better, if that's not too much to ask.

Even Haskellers use QuickCheck :)